### PR TITLE
add types for stores

### PR DESCRIPTION
**Pull request by TheComputerM**
Solves https://github.com/sveltejs/sapper/issues/1556
### Before submitting the PR, please make sure you do the following
- [x] It's really useful if your PR references an issue where it is discussed ahead of time. In many cases, features are absent for a reason. For large changes, please create an RFC: https://github.com/sveltejs/rfcs
- [x] This message body should clearly illustrate what problems it solves.
- [ ] Ideally, include a test that fails without this PR but passes with it.

### Tests
-  [x] Run the tests with `npm test` and lint the project with `npm run lint`

### CODE REVIEW

**Comment by benmccann** [comment]
Please change the formatting back to match the original (e.g. tabs instead of spaces)

**Comment by TheComputerM** [comment]
Shall I also add a `.editorconfig` to enforce formatting in the editor?

**Comment by TheComputerM** [comment]
Please inform me if you want me to add more type definitions.

File: runtime/index.d.ts [hunk 1]
```diff
@@ -1,28 +1,53 @@
-declare module "@sapper/app"
-declare module "@sapper/server"
-declare module "@sapper/service-worker"
+declare module "@sapper/app";
+declare module "@sapper/server";
+declare module "@sapper/service-worker";
+
+interface Stores {
+	preloading: {
+		subscribe(preloading: boolean): () => void;
```
Comment (ehrencrona) [inline runtime/index.d.ts:7]: There are ready-made types for Stores in svelte that you could reuse that give a more complete type definition for subscribe (there's an optional second parameter and a return value that are missing here):

I think if you just do

```
import { Readable, Writable } from 'svelte/store';

...
	preloading: Readable<boolean>;
```

you should get the correct definitions for `subscribe` etc for free.

For the page store, there is also already a `PageContext` interface, so you could declare it as

```
	import { PageContext } from '@sapper/app/types';
...
	page: Writeable<PageContext>;
```

**Review by ehrencrona — CHANGES_REQUESTED** [review]

File: runtime/index.d.ts (hunk 1 above)
Comment (TheComputerM) [reply runtime/index.d.ts]: Done

**Comment by benmccann** [comment]
@TheComputerM this PR will need to be rebased

**Comment by TheComputerM** [comment]
Sorry, I had forgotten this.

File: runtime/index.d.ts [hunk 2]
```diff
@@ -1,24 +1,40 @@
-declare module '@sapper/app' {
+import { Readable, Writable } from "svelte/store";
+import { PageContext } from "@sapper/app/types";
+
+declare module "@sapper/app";
```
Comment (benmccann) [inline runtime/index.d.ts:4]: there were just removed in https://github.com/sveltejs/sapper/pull/1598. you might need to rebase against `master`

File: runtime/index.d.ts [hunk 3]
```diff
@@ -1,24 +1,40 @@
...
+		location: string;
 	}
 
-	export function goto(href: string, opts: { noscroll?: boolean, replaceState?: boolean }): Promise<void>;
-	export function prefetch(href: string): Promise<{ redirect?: Redirect; data?: unknown }>;
+	export function goto(
+		href: string,
+		opts: { noscroll?: boolean; replaceState?: boolean }
+	): Promise<void>;
+	export function prefetch(
```
Comment (benmccann) [inline runtime/index.d.ts:18]: can you leave the method formatting the way it was?

**Comment by benmccann** [comment]
The linting and tests are failing with this PR

**Comment by TheComputerM** [comment]
I am going to open a new PR because this is getting cluttered.

**Comment by TheComputerM** [comment]
New PR: https://github.com/sveltejs/sapper/pull/1601

**Pull request closed without merging**